### PR TITLE
Save topk models

### PR DESCRIPTION
**Pull request by samadejacobs**
Save  topk models from a multi-trainer training for inference/post processing.

### CODE REVIEW

File: src/callbacks/callback_save_topk_models.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  else  std::sort(score_v.begin(), score_v.end(),std::greater<EvalType>());
+  score_v.resize(m_k);
+  
+  if (comm->am_world_master()) {
+    std::cout << "Top " << m_k << " " << m_metric_name << " average "
+              << std::accumulate(score_v.begin(), score_v.end(), EvalType(0))/m_k << std::endl;
+  } 
+  for(int i =0; i < num_trainers; ++i) {
+    if(std::find(score_v.begin(), score_v.end(), 
+                 score_list[i]) != score_v.end()) {
```
Comment (timmoon10) [inline src/callbacks/callback_save_topk_models.cpp]: If multiple trainers have the same score, more than k trainers may be saved. In the extreme case, every trainer will be saved if all trainers have the same score. If we want to handle this edge case, we need to do the sorting more carefully.

If we don't mind the edge case and leave it as it is, we don't need to iterate through all the trainers. We only need to check if this trainer's score is in the top-k.

File: src/callbacks/callback_save_topk_models.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  
+  if (comm->am_world_master()) {
+    std::cout << "Top " << m_k << " " << m_metric_name << " average "
+              << std::accumulate(score_v.begin(), score_v.end(), EvalType(0))/m_k << std::endl;
+  } 
+  for(int i =0; i < num_trainers; ++i) {
+    if(std::find(score_v.begin(), score_v.end(), 
+                 score_list[i]) != score_v.end()) { 
+      if( i == comm->get_trainer_rank()) {
+        std::cout << "Trainer [ " << comm->get_trainer_rank() << "] in top list with score " << score_list[i] << std::endl;;
```
[the remark this answers is not in the document]
Comment (samadejacobs) [reply src/callbacks/callback_save_topk_models.cpp]: Done, too much verbosity!

File: include/lbann/callbacks/callback_save_topk_models.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  lbann_callback_save_topk_models(std::string dir, int k, std::string metric_name, bool ascending_ordering=false) : 
+  lbann_callback_save_model(dir,true), m_k(k),m_metric_name(metric_name),m_ascending_ordering(ascending_ordering) {}
+  lbann_callback_save_topk_models(const lbann_callback_save_topk_models&) = default;
+  lbann_callback_save_topk_models& operator=(const lbann_callback_save_topk_models&) = default;
+  lbann_callback_save_topk_models* copy() const override { return new lbann_callback_save_topk_models(*this); }
+  void on_test_end(model *m) override;
+  std::string name() const override { return "save_topk_models"; }
+
+ private:
+  bool compute_stats(model *m);
```
[the remark this answers is not in the document]
Comment (samadejacobs) [reply include/lbann/callbacks/callback_save_topk_models.hpp]: Done

File: src/callbacks/callback_save_topk_models.cpp (hunk 1 above)
Comment (samadejacobs) [reply src/callbacks/callback_save_topk_models.cpp]: The evaluation metric is just the first pass, so it is ok to save all the "good models" that meet initial condition for post-processing/ further evaluation (e.g., physics constraints)

**Comment by timmoon10** [comment]
It would be good to document that this may end up saving more than `k` models if multiple have the same metric score.

**Pull request merged**